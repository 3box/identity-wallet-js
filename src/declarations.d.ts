// not sure why this is needed
declare module 'ipfs' {
  export default any
}

declare module 'multihashes' {
  export function encode(digest: Buffer, code: number | string, length?: number): Buffer
}
