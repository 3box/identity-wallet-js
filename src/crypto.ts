import { randomBytes } from '@stablelib/random'
export { randomBytes }
import {
  createJWE,
  decryptJWE,
  JWE,
  Encrypter,
  Decrypter,
  x25519Encrypter,
  x25519Decrypter,
} from 'did-jwt'
import { prepareCleartext, decodeCleartext } from 'dag-jose-utils'

interface EncryptOpts {
  publicKey?: Uint8Array
  encrypter?: Encrypter
  kid?: string
}

function isU8a(b: Uint8Array | undefined): b is Uint8Array {
  return b instanceof Uint8Array
}

export async function asymEncryptJWE(
  cleartext: Record<string, any>,
  { publicKey, encrypter, kid }: EncryptOpts
): Promise<JWE> {
  if (!encrypter) {
    if (!isU8a(publicKey)) throw new Error('publicKey or encrypter has to be defined')
    encrypter = x25519Encrypter(publicKey, kid)
  }
  return createJWE(prepareCleartext(cleartext), [encrypter])
}

interface DecryptOpts {
  secretKey?: Uint8Array
  decrypter?: Decrypter
}

export async function asymDecryptJWE(
  jwe: JWE,
  { secretKey, decrypter }: DecryptOpts
): Promise<Record<string, any>> {
  if (!decrypter) {
    if (!isU8a(secretKey)) throw new Error('secretKey or decrypter has to be defined')
    decrypter = x25519Decrypter(secretKey)
  }
  return decodeCleartext(await decryptJWE(jwe, decrypter))
}

interface Recipient {
  header: Record<string, string>
}

export function parseJWEKids(jwe: JWE): Array<string> {
  return (jwe.recipients || []).reduce((kids: Array<string>, recipient: Recipient) => {
    if (recipient.header?.kid) kids.push(recipient.header.kid.split('#')[1])
    return kids
  }, [])
}
