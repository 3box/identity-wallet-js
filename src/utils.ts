import { sha256 } from 'js-sha256'
import Multihash from 'multihashes'
import naclutil from 'tweetnacl-util'
import bs58 from 'bs58'
import dagCBOR from 'ipld-dag-cbor'
import { Wallet } from '@ethersproject/wallet'
import stringify from 'fast-json-stable-stringify'

const ENC_BLOCK_SIZE = 24

export const pad = (val: string, blockSize = ENC_BLOCK_SIZE): string => {
  const blockDiff = (blockSize - (val.length % blockSize)) % blockSize
  return `${val}${'\0'.repeat(blockDiff)}`
}

export const unpad = (padded: string): string => padded.replace(/\0+$/, '')

export const sha256Multihash = (str: string): string => {
  return Multihash.encode(Buffer.from(sha256(str)), 'sha2-256').toString('hex')
}

const multicodecPubkeyTable = {
  'secp256k1': 0xe7,
  'x25519': 0xec,
  'ed25519': 0xed
}

function encodeKey(key: Uint8Array, keyType: string) {
  const buf = new Uint8Array(key.length + 2)
  if (!multicodecPubkeyTable[keyType]) {
    throw new Error(`Key type "${keyType}" not supported.`)
  }
  buf[0] = multicodecPubkeyTable[keyType]
  // The multicodec is encoded as a varint so we need to add this.
  // See js-multicodec for a general implementation
  buf[1] = 0x01
  buf.set(key, 2)
  return `z${bs58.encode(buf)}`
}

export const gen3IDgenesis = (pubkeys: any): any => {
  return {
    metadata: {
      owners: [encodeKey(hexToUint8Array(pubkeys.managementKey), 'secp256k1')],
    },
    content: {
      publicKeys: {
        signing: encodeKey(hexToUint8Array(pubkeys.signingKey), 'secp256k1'),
        encryption: encodeKey(naclutil.decodeBase64(pubkeys.asymEncryptionKey), 'x25519')
      }
    }
  }
}

let tmpData: any
export const fakeIpfs = {
  dag: {
    put: (data: any, opts: dagCBOR.UserOptions) => {
      tmpData = data
      return dagCBOR.util.cid(dagCBOR.util.serialize(data), opts)
    },
    get: () => ({ value: tmpData }),
  },
  add: () => 'empty', // used in _initMuport in 3box-js 3id, but muport fingerprint not needed here
}

export const fakeEthProvider = (wallet: Wallet) => ({
  send: (
    request: { method: string; params: Array<any> },
    callback: (err: Error | null | undefined, res?: any) => void
  ) => {
    if (request.method !== 'personal_sign') {
      callback(new Error('only supports personal_sign'))
    } else {
      let message = request.params[0]
      if (message.startsWith('0x')) {
        message = Buffer.from(message.slice(2), 'hex').toString('utf8')
      }
      callback(null, { result: wallet.signMessage(message) })
    }
  },
})

export function hexToUint8Array(str: string): Uint8Array {
  return new Uint8Array(Buffer.from(str, 'hex'))
}

export function toStableObject(obj: any): any {
  return JSON.parse(stringify(obj))
}
