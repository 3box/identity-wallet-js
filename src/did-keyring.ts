import nacl, { BoxKeyPair } from 'tweetnacl'
import naclutil from 'tweetnacl-util'
import { HDNode } from '@ethersproject/hdnode'
import { Wallet } from '@ethersproject/wallet'
import { SimpleSigner } from 'did-jwt'
import { sha256 } from 'js-sha256'
import { ec as EC } from 'elliptic'

import { naclRandom } from './crypto'

const ec = new EC('secp256k1')

const BASE_PATH = "m/51073068'/0'"
const ROOT_STORE_PATH = "0'/0'/0'/0'/0'/0'/0'/0'"

const AUTH_PATH_WALLET = BASE_PATH + '/' + ROOT_STORE_PATH + '/0'
const AUTH_PATH_ENCRYPTION = BASE_PATH + '/' + ROOT_STORE_PATH + '/3'

const ensure0x = (str: string): string => {
  return (str.startsWith('0x') ? '' : '0x') + str
}

interface KeysSet {
  signingKey: HDNode
  asymEncryptionKey: BoxKeyPair
  symEncryptionKey: Uint8Array
  managementKey: HDNode
}

export class Keyring {
  public static create() {
    const seed = '0x' + Buffer.from(naclRandom(32)).toString('hex')
    return new Keyring(seed)
  }

  private _seed: string
  private _baseNode: HDNode
  private _keys: KeysSet

  constructor(seed: string) {
    this._seed = seed
    this._baseNode = HDNode.fromSeed(seed).derivePath(BASE_PATH)
    const rootNode = this._baseNode.derivePath(ROOT_STORE_PATH)
    this._keys = {
      signingKey: rootNode.derivePath('0'),
      asymEncryptionKey: nacl.box.keyPair.fromSecretKey(
        new Uint8Array(
          Buffer.from(rootNode.derivePath('2').privateKey.slice(2), 'hex'),
        ),
      ),
      symEncryptionKey: Keyring.hexToUint8Array(
        rootNode.derivePath('3').privateKey.slice(2),
      ),
      managementKey: rootNode.derivePath('1'),
    }
  }

  _getKeys() {
    return this._keys
  }

  asymEncrypt(msg, toPublic, { nonce } = {}) {
    nonce = nonce || Keyring.randomNonce()
    toPublic = naclutil.decodeBase64(toPublic)
    if (typeof msg === 'string') {
      msg = naclutil.decodeUTF8(msg)
    }
    const ephemneralKeypair = nacl.box.keyPair()
    const ciphertext = nacl.box(
      msg,
      nonce,
      toPublic,
      ephemneralKeypair.secretKey,
    )
    return {
      nonce: naclutil.encodeBase64(nonce),
      ephemeralFrom: naclutil.encodeBase64(ephemneralKeypair.publicKey),
      ciphertext: naclutil.encodeBase64(ciphertext),
    }
  }

  asymDecrypt(ciphertext, fromPublic, nonce, { space, toBuffer } = {}) {
    fromPublic = naclutil.decodeBase64(fromPublic)
    ciphertext = naclutil.decodeBase64(ciphertext)
    nonce = naclutil.decodeBase64(nonce)
    const cleartext = nacl.box.open(
      ciphertext,
      nonce,
      fromPublic,
      this._keys.asymEncryptionKey.secretKey,
    )
    if (toBuffer) {
      return cleartext ? Buffer.from(cleartext) : null
    }
    return cleartext ? nacl.util.encodeUTF8(cleartext) : null
  }

  symEncrypt(msg, { space, nonce } = {}) {
    return Keyring.symEncryptBase(msg, this._keys.symEncryptionKey, nonce)
  }

  symDecrypt(ciphertext, nonce, { space, toBuffer } = {}) {
    return Keyring.symDecryptBase(
      ciphertext,
      this._keys.symEncryptionKey,
      nonce,
      toBuffer,
    )
  }

  managementPersonalSign(message) {
    const wallet = this.managementWallet()
    return wallet.signMessage(message)
  }

  managementWallet() {
    return new Wallet(this._keys.managementKey.privateKey)
  }

  getJWTSigner() {
    return SimpleSigner(this._keys.signingKey.privateKey.slice(2))
  }

  getDBSalt() {
    return sha256(this._keys.signingKey.derivePath('0').privateKey.slice(2))
  }

  getPublicKeys({ uncompressed }) {
    let signingKey = this._keys.signingKey.publicKey.slice(2)
    const managementKey = this._keys.managementKey.address
    if (uncompressed) {
      signingKey = ec
        .keyFromPublic(Buffer.from(signingKey, 'hex'))
        .getPublic(false, 'hex')
    }
    return {
      signingKey,
      managementKey,
      asymEncryptionKey: naclutil.encodeBase64(
        this._keys.asymEncryptionKey.publicKey,
      ),
    }
  }

  serialize() {
    return this._seed
  }

  static encryptWithAuthSecret(message, authSecret) {
    const node = HDNode.fromSeed(ensure0x(authSecret)).derivePath(
      AUTH_PATH_ENCRYPTION,
    )
    const key = Keyring.hexToUint8Array(node.privateKey.slice(2))
    return Keyring.symEncryptBase(message, key)
  }

  static decryptWithAuthSecret(ciphertext, nonce, authSecret) {
    const node = HDNode.fromSeed(ensure0x(authSecret)).derivePath(
      AUTH_PATH_ENCRYPTION,
    )
    const key = Keyring.hexToUint8Array(node.privateKey.slice(2))
    return Keyring.symDecryptBase(ciphertext, key, nonce)
  }

  static walletForAuthSecret(authSecret) {
    const node = HDNode.fromSeed(ensure0x(authSecret)).derivePath(
      AUTH_PATH_WALLET,
    )
    return new Wallet(node.privateKey)
  }

  static hexToUint8Array(str) {
    return new Uint8Array(Buffer.from(str, 'hex'))
  }

  static symEncryptBase(msg, symKey, nonce) {
    nonce = nonce || Keyring.randomNonce()
    if (typeof msg === 'string') {
      msg = nacl.util.decodeUTF8(msg)
    }
    const ciphertext = nacl.secretbox(msg, nonce, symKey)
    return {
      nonce: nacl.util.encodeBase64(nonce),
      ciphertext: nacl.util.encodeBase64(ciphertext),
    }
  }

  static symDecryptBase(ciphertext, symKey, nonce, toBuffer) {
    ciphertext = nacl.util.decodeBase64(ciphertext)
    nonce = nacl.util.decodeBase64(nonce)
    const cleartext = nacl.secretbox.open(ciphertext, nonce, symKey)
    if (toBuffer) {
      return cleartext ? Buffer.from(cleartext) : null
    }
    return cleartext ? nacl.util.encodeUTF8(cleartext) : null
  }

  static naclRandom(length) {
    return nacl.randomBytes(length)
  }

  static randomNonce() {
    return Keyring.naclRandom(24)
  }
}

module.exports = Keyring
