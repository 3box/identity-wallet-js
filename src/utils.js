import crypto from 'crypto'
import Multihash from 'multihashes'

const ENC_BLOCK_SIZE = 24

const pad = (val, blockSize = ENC_BLOCK_SIZE) => {
  const blockDiff = (blockSize - (val.length % blockSize)) % blockSize
  return `${val}${'\0'.repeat(blockDiff)}`
}

const unpad = padded => padded.replace(/\0+$/, '')

const sha256 = msg => crypto.createHash('sha256').update(msg).digest('hex')

const sha256Multihash = str => Multihash.encode(Buffer.from(sha256(str)), 'sha2-256').toString('hex')

module.exports = {
  pad,
  unpad,
  sha256,
  sha256Multihash
}
