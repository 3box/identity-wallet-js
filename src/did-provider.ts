import { createJWS } from 'did-jwt'
import {
  HandlerMethods,
  RequestHandler,
  RPCConnection,
  RPCError,
  RPCRequest,
  RPCResponse,
  createHandler,
} from 'rpc-utils'

import Keyring from './keyring'
import { ThreeIDX } from './three-idx'
import Permissions from './permissions'
import { toStableObject } from './utils'

type Origin = string | null | undefined

export type Context = {
  permissions: Permissions
  threeIdx: ThreeIDX
  keyring: Keyring
  origin: Origin
}

interface CreateJWSParams {
  payload: Record<string, any>
  protected?: Record<string, any>
  did: string
}

interface AuthParams {
  paths: Array<string>
}

export const didMethods: HandlerMethods<Context> = {
  did_authenticate: async ({ permissions, threeIdx, origin }, params: AuthParams) => {
    const paths = await permissions.request(origin, params.paths || [])
    // paths should be an array if permission granted
    // may be a subset or requested paths or empty array
    if (paths === null) throw new RPCError(4001, 'User Rejected Request')
    return { did: threeIdx.DID, paths }
  },
  did_createJWS: async ({ permissions, keyring, threeIdx, origin }, params: CreateJWSParams) => {
    if (!permissions.has(origin)) {
      throw new RPCError(4100, 'Unauthorized')
    }
    // TODO - if the requesting DID is our management key
    // (did:key) we should request explicit permission.
    const keyName = threeIdx.parseKeyName(params.did)
    const kid = await threeIdx.encodeKidWithVersion(keyName)
    const signer = keyring.getSigner(keyName)
    const header = toStableObject(Object.assign(params.protected || {}, { kid }))
    const jws = await createJWS(toStableObject(params.payload), signer, header)
    return { jws }
  },
}

export interface ProviderConfig {
  permissions: Permissions
  threeIdx: ThreeIDX
  keyring: Keyring
  forcedOrigin?: string
}

export class DidProvider implements RPCConnection {
  protected _handle: RequestHandler

  constructor({ permissions, threeIdx, keyring, forcedOrigin }: ProviderConfig) {
    const handler = createHandler<Context>(didMethods)
    this._handle = (origin: string, msg: RPCRequest) => {
      return handler(
        {
          origin: forcedOrigin || origin,
          permissions,
          threeIdx,
          keyring,
        },
        msg
      )
    }
  }

  public get isDidProvider(): boolean {
    return true
  }

  public async send(msg: RPCRequest, origin?: Origin): Promise<RPCResponse | null> {
    return await this._handle(origin, msg)
  }
}
