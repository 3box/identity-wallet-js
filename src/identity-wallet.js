import EventEmitter from 'events'
import Keyring from './keyring'
import didJWT from 'did-jwt'
import { sha256Multihash, pad, unpad } from './utils'


class IdentityWallet {

  /**
   * Creates an instance of IdentityWallet
   *
   * @param     {Object}    config                  The configuration to be used
   * @param     {String}    config.seed             The seed of the identity, 32 hex string
   * @param     {String}    config.authSecret       The authSecret to use, 32 hex string
   * @param     {String}    config.ethereumAddress  The ethereumAddress of the identity
   * @return    {this}                            An IdentityWallet instance
   */
  constructor (config = {}) {
    this.events = new EventEmitter()
    if (config.seed) {
      this._keyring = new Keyring(config.seed)
    }
    if (config.authSecret) {
      if (!config.ethereumAddress) throw new Error('Ethereum address needs to be defined when authSecret given')
      this._authSecret = config.authSecret
      this._ethereumAddress = config.ethereumAddress
    }
  }

  async getAddress () {
    return this._ethereumAddress
  }

  _initKeyring (authData) {
    let seed
    if (authData) {
      authData.find(({ ciphertext, nonce }) => {
        const key = Keyring.hexToUint8Array(this._authSecret)
        seed = Keyring.symDecryptBase(ciphertext, key, nonce)
        return Boolean(seed)
      })
      if (!seed) throw new Error('No valid auth-secret for this identity')
    } else {
      // no authData available so we create a new identity
      seed = '0x' + Buffer.from(Keyring.naclRandom(32)).toString('hex')
      this.addAuthMethod(this._authSecret, seed)
    }
    this._keyring = new Keyring(seed)
  }

  /**
   * Authenticate to given spaces
   *
   * @param     {Array<String>}     spaces          The desired spaces
   * @param     {Object}    opts                    Optional parameters
   * @param     {Array<Object>}     opts.authData   The authData for this identity
   * @return    {Object}                            The public keys for the requested spaces of this identity
   */
  async authenticate (spaces = [], { authData } = {}) {
    if (!this._keyring) this._initKeyring(authData)
    const result = {
      main: this._keyring.getPublicKeys(),
      spaces: {}
    }
    if (this._ethereumAddress) result.main.managementKey = this._ethereumAddress
    spaces.map(space => {
      result.spaces[space] = this._keyring.getPublicKeys({ space, uncompressed: true })
    })
    return result
  }

  /**
   * Check if authenticated to given spaces
   *
   * @param     {Array<String>}     spaces          The desired spaces
   * @return    {Boolean}                           True if authenticated
   */
  async isAuthenticated (spaces = []) {
    return Boolean(this._keyring)
  }

  /**
   * Add a new authentication method for this identity
   *
   * @param     {String}    authSecret              A 32 byte hex string used as authentication secret
   */
  async addAuthMethod (authSecret, seed) {
    if (!seed && !this._keyring) throw new Error('This method can only be called after authenticate has been called')

    const message = seed || this._keyring._seed
    const key = Keyring.hexToUint8Array(authSecret)
    const encAuthData = Keyring.symEncryptBase(message, key)

    this.events.emit('new-auth-method', encAuthData)
  }

  /**
   * Sign a verifiable credential. The format of the credential is [did-jwt](https://github.com/uport-project/did-jwt).
   *
   * @param     {Object}    payload                 The payload of the claim
   * @param     {Object}    opts                    Optional parameters
   * @param     {String}    opts.DID                The DID used as the issuer of this claim
   * @param     {String}    opts.space              The space used to sign the claim
   * @param     {String}    opts.expiresIn          Set an expiry date for the claim as unix timestamp
   * @return    {String}                            The signed claim encoded as a JWT
   */
  async signClaim (payload, { DID, space, expiresIn } = {}) {
    if (!this._keyring) throw new Error('This method can only be called after authenticate has been called')

    const issuer = DID
    const settings = {
      signer: this._keyring.getJWTSigner(space),
      issuer,
      expiresIn
    }
    return didJWT.createJWT(payload, settings)
  }

  /**
   * Encrypt a message
   *
   * @param     {String}    message                 The message to be encrypted
   * @param     {String}    space                   The space used for encryption
   * @param     {Object}    opts                    Optional parameters
   * @param     {String}    opts.nonce              The nonce used to encrypt the message
   * @param     {String}    opts.blockSize          The blockSize used for padding (default 24)
   * @return    {Object}                            The encrypted object (ciphertext and nonce)
   */
  async encrypt (message, space, { nonce, blockSize } = {}) {
    if (!this._keyring) throw new Error('This method can only be called after authenticate has been called')

    const paddedMsg = pad(message, blockSize)
    return this._keyring.symEncrypt(paddedMsg, { space, nonce })
  }

  /**
   * Decrypt a message
   *
   * @param     {Object}    encryptedObject         The encrypted object (ciphertext and nonce)
   * @param     {String}    space                   The space used for encryption
   * @return    {String}                            The decrypted message
   */
  async decrypt (encObj, space) {
    if (!this._keyring) throw new Error('This method can only be called after authenticate has been called')

    const paddedMsg = this._keyring.symDecrypt(encObj.ciphertext, encObj.nonce, { space })
    if (!paddedMsg) throw new Error('IdentityWallet: Could not decrypt message')
    return unpad(paddedMsg)
  }

  async hashDBKey(key, space) {
    const salt = this._keyring.getDBSalt(space)
    return sha256Multihash(salt + key)
  }
}


module.exports = IdentityWallet
