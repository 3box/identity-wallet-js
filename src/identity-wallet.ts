import DidDocument from 'ipfs-did-document'
import { LinkProof, createLink } from '3id-blockchain-utils'
import { CeramicApi } from '@ceramicnetwork/ceramic-common'

import { AsymEncryptedMessage, EncryptedMessage, naclRandom } from './crypto'
import { DidProvider } from './did-provider'
import Keyring, { PublicKeys } from './keyring'
import ThreeIdProvider from './threeIdProvider'
import { ThreeIDX } from './three-idx'
import { sha256Multihash, pad, unpad, fakeIpfs, fakeEthProvider } from './utils'

import Permissions, { GetPermissionFn, SELF_ORIGIN } from './permissions'

const DID_METHOD_NAME = '3'

interface IDWConfig {
  getPermission: GetPermissionFn
  seed?: string
  authSecret?: string
  externalAuth?: (req: any) => Promise<any>
  ceramic: CeramicApi
  useThreeIdProv: boolean
}

export default class IdentityWallet {
  protected _seed: string | undefined
  protected _authSecret: string | undefined
  protected _externalAuth: ((req: any) => Promise<any>) | undefined
  protected _keyring: Keyring | undefined

  public DID: string | undefined

  constructor(config: IDWConfig) {
    this.permissions = new Permissions(config.getPermission)
    if (config.seed) {
      this._seed = config.seed
    } else if (config.authSecret) {
      this._authSecret = config.authSecret
    } else if (config.externalAuth) {
      this._externalAuth = config.externalAuth
    } else {
      throw new Error(
        'Either seed, or authSecret has to be passed to create an IdentityWallet instance'
      )
    }
  }

  /**
   * Creates an instance of IdentityWallet
   *
   * @param     {Object}    config                  The configuration to be used
   * @param     {Function}  config.getPermission    The function that is called to ask the user for permission
   * @param     {String}    config.seed             The seed of the identity, 32 hex string
   * @param     {String}    config.authSecret       The authSecret to use, 32 hex string
   * @param     {String}    config.externalAuth     External auth function, directly returns key material, used to migrate legacy 3box accounts
   * @return    {IdentityWallet}                    An IdentityWallet instance
   */
  static async create(config: IDWConfig) {
    // the next two lines will likely change soon
    const idw = new IdentityWallet(config)
    idw._init(config.ceramic)
  }

  /**
   * Temporary option for use with ceramic js-did migration
   */
  async _init(ceramic) {
    if (!this._seed) throw new Error('seed required for now')
    this._keyring = new Keyring(this._seed)
    this.threeIdx = new ThreeIDX(ceramic)
    const pubkeys = this._keyring.getPublicKeys({ mgmtPub: true })
    await this.threeIdx.create3idDoc(pubkeys)
    // TODO - change to DID provider when ceramic uses js-did
    await this.threeIdx.setDIDProvider(this.get3idProvider(SELF_ORIGIN))
  }

  /**
   * Get the 3IDProvider
   *
   * @return    {ThreeIdProvider}                   The 3IDProvider for this IdentityWallet instance
   */
  get3idProvider(forcedOrigin?: string) {
    return new ThreeIdProvider(this)
  }

  /**
   * Get the DIDProvider
   *
   * @return    {DidProvider}                   The DIDProvider for this IdentityWallet instance
   */
  getDidProvider(forcedOrigin?: string) {
    return new DidProvider({
      keyring: this._keyring,
      permissions: this.permissions,
      threeIdx: this.threeIdx,
      forcedOrigin
    })
  }
}
