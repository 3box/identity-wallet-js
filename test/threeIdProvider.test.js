import ThreeIdProvider from '../src/threeIdProvider'
import { sha256Multihash, pad, unpad } from '../src/utils'

function formatCall(method, params) {
  return {
    id: 1,
    jsonrpc: '2.0',
    method: `3id_${method}`,
    params,
  }
}

async function callWithCB(rpc, payload, origin) {
  return new Promise((resolve, reject) => {
    const callback = (err, resp) => {
      if (err) reject(err)
      else resolve(resp)
    }
    if (origin) {
      rpc.send(payload, origin, callback)
    } else {
      rpc.send(payload, callback)
    }
  })
}

describe('ThreeIdProvider', () => {
  let rpc

  it('is 3id provider', async () => {
    rpc = new ThreeIdProvider({})
    expect(rpc.is3idProvider).toBeTruthy()
    expect(rpc.isDidProvider).toBeTruthy()
  })

  it('authenticate correctly', async () => {
    const config = {
      permissions: { request: jest.fn(async (o, spaces) => spaces) },
      keyring: { getPublicKeys: jest.fn(({ space }) => {
        return space ? 'spacekeys' : 'mainkeys'
      })}
    }
    rpc = new ThreeIdProvider(config)
    const origin = 'https://my.origin'
    const spaces = ['space1']
    const authData = ['enc auth data']
    const payload = formatCall('authenticate', { spaces, authData })
    expect(await rpc.send(payload, origin)).toMatchSnapshot()
    expect(config.keyring.getPublicKeys).toHaveBeenCalledTimes(2)
    expect(config.keyring.getPublicKeys).toHaveBeenNthCalledWith(1, {})
    expect(config.keyring.getPublicKeys).toHaveBeenNthCalledWith(2, { space: spaces[0], uncompressed: true })
    const payload2 = formatCall('authenticate', { spaces, authData, mgmtPub: true })
    expect(await rpc.send(payload2, origin)).toMatchSnapshot()
    expect(config.keyring.getPublicKeys).toHaveBeenCalledTimes(4)
    expect(config.keyring.getPublicKeys).toHaveBeenNthCalledWith(3, { mgmtPub: true })
    expect(config.keyring.getPublicKeys).toHaveBeenNthCalledWith(4, { space: spaces[0], uncompressed: true })
  })

  it('isAuthenticated correctly', async () => {
    const config = {
      permissions: { has: jest.fn(() => true) },
    }
    rpc = new ThreeIdProvider(config)
    const origin = 'https://my.origin'
    const spaces = ['space1']
    const payload = formatCall('isAuthenticated', { spaces })
    expect(await rpc.send(payload, origin)).toMatchSnapshot()
    expect(config.permissions.has).toHaveBeenCalledTimes(1)
    expect(config.permissions.has).toHaveBeenCalledWith(origin, spaces)
  })

  it('sign claim correctly', async () => {
    const config = {
      permissions: { has: jest.fn(() => true) },
      threeIdx: { DID: 'did:3:asdf' },
      keyring: { getJWTSigner: () => () => Promise.resolve('signed') },
    }
    rpc = new ThreeIdProvider(config)
    const pl = { d: 'data', iat: undefined }
    const space = 'space1'
    const payload = formatCall('signClaim', { payload: pl, space })
    expect(await rpc.send(payload)).toMatchSnapshot()
    expect(await callWithCB(rpc, payload)).toMatchSnapshot()
  })

  it('encrypt correctly', async () => {
    const config = {
      permissions: { has: jest.fn(() => true) },
      keyring: { symEncrypt: jest.fn(() => 'encrypted msg') }
    }
    rpc = new ThreeIdProvider(config)
    const message = 'data'
    const space = 'space1'
    const payload = formatCall('encrypt', { message, space })
    expect((await rpc.send(payload)).result).toEqual('encrypted msg')
    expect(config.keyring.symEncrypt).toHaveBeenCalledTimes(1)
    expect(config.keyring.symEncrypt).toHaveBeenCalledWith(pad(message), { space })
  })

  it('asymmetrically encrypt correctly', async () => {
    const config = {
      permissions: { has: jest.fn(() => true) },
      keyring: { asymEncrypt: jest.fn(() => 'encrypted msg') }
    }
    rpc = new ThreeIdProvider(config)
    const message = 'data'
    const to = 'pubkey'
    const payload = formatCall('encrypt', { message, to })
    expect((await rpc.send(payload)).result).toEqual('encrypted msg')
    expect(config.keyring.asymEncrypt).toHaveBeenCalledTimes(1)
    expect(config.keyring.asymEncrypt).toHaveBeenCalledWith(pad(message), to)
  })

  it('decrypt correctly', async () => {
    const config = {
      permissions: { has: jest.fn(() => true) },
      keyring: { symDecrypt: jest.fn(() => 'decrypted msg') }
    }
    rpc = new ThreeIdProvider(config)
    const ciphertext = 'data'
    const nonce = 'nonce'
    const space = 'space1'
    const payload = formatCall('decrypt', { ciphertext, nonce, space })
    expect(await rpc.send(payload)).toMatchSnapshot()
    expect(config.keyring.symDecrypt).toHaveBeenCalledTimes(1)
    expect(config.keyring.symDecrypt).toHaveBeenCalledWith(ciphertext, nonce, { space })
  })

  it('asymmetrically decrypt correctly', async () => {
    const config = {
      permissions: { has: jest.fn(() => true) },
      keyring: { asymDecrypt: jest.fn(() => 'decrypted msg') }
    }
    rpc = new ThreeIdProvider(config)
    const ciphertext = 'data'
    const nonce = 'nonce'
    const space = 'space1'
    const ephemeralFrom = 'publicKey'
    const payload = formatCall('decrypt', {
      ciphertext,
      nonce,
      space,
      ephemeralFrom,
    })
    expect(await rpc.send(payload)).toMatchSnapshot()
    expect(config.keyring.asymDecrypt).toHaveBeenCalledTimes(1)
    expect(config.keyring.asymDecrypt).toHaveBeenCalledWith(
      ciphertext, ephemeralFrom, nonce, { space }
    )
  })

  it('hash entry key correctly', async () => {
    const config = {
      permissions: { has: jest.fn(() => true) },
      keyring: { getDBSalt: jest.fn(() => 'saltysalt') }
    }
    rpc = new ThreeIdProvider(config)
    const key = 'key'
    const space = 'space1'
    const payload = formatCall('hashEntryKey', { key, space })
    expect((await rpc.send(payload)).result).toEqual(sha256Multihash('saltysalt' + key))
    expect(config.keyring.getDBSalt).toHaveBeenCalledTimes(1)
    expect(config.keyring.getDBSalt).toHaveBeenCalledWith(space)
  })

  it('Unsupported method should throw', async () => {
    const payload = formatCall('notSupported')
    await expect(rpc.send(payload)).rejects.toMatchSnapshot()
    await expect(callWithCB(rpc, payload)).rejects.toMatchSnapshot()
  })
})
