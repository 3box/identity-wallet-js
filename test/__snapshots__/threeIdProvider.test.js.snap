// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeIdProvider Unsupported method should throw 1`] = `[Error: Method not found]`;

exports[`ThreeIdProvider Unsupported method should throw 2`] = `[Error: Method not found]`;

exports[`ThreeIdProvider asymmetrically decrypt correctly 1`] = `
Object {
  "id": 1,
  "jsonrpc": "2.0",
  "result": "decrypted msg",
}
`;

exports[`ThreeIdProvider authenticate correctly 1`] = `
Object {
  "id": 1,
  "jsonrpc": "2.0",
  "result": Object {
    "main": "mainkeys",
    "spaces": Object {
      "space1": "spacekeys",
    },
  },
}
`;

exports[`ThreeIdProvider authenticate correctly 2`] = `
Object {
  "id": 1,
  "jsonrpc": "2.0",
  "result": Object {
    "main": "mainkeys",
    "spaces": Object {
      "space1": "spacekeys",
    },
  },
}
`;

exports[`ThreeIdProvider decrypt correctly 1`] = `
Object {
  "id": 1,
  "jsonrpc": "2.0",
  "result": "decrypted msg",
}
`;

exports[`ThreeIdProvider isAuthenticated correctly 1`] = `
Object {
  "id": 1,
  "jsonrpc": "2.0",
  "result": true,
}
`;

exports[`ThreeIdProvider sign claim correctly 1`] = `
Object {
  "id": 1,
  "jsonrpc": "2.0",
  "result": "eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NksifQ.eyJkIjoiZGF0YSIsImlzcyI6ImRpZDozOmFzZGYifQ.signed",
}
`;

exports[`ThreeIdProvider sign claim correctly 2`] = `
Object {
  "id": 1,
  "jsonrpc": "2.0",
  "result": "eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NksifQ.eyJkIjoiZGF0YSIsImlzcyI6ImRpZDozOmFzZGYifQ.signed",
}
`;
